src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -11302,15 +11320,55 @@ private bool ScanCast(bool forPattern = false)
 
         private bool IsPossibleLambdaExpression(Precedence precedence)
         {
-            // Only call into this if after `static` or after a legal identifier.
+            // Only call into this if after `static`, '[',  or after a legal identifier.
             Debug.Assert(
-                this.CurrentToken.Kind == SyntaxKind.StaticKeyword ||
+                this.CurrentToken.Kind is SyntaxKind.StaticKeyword or SyntaxKind.OpenBracketToken ||
                 this.IsTrueIdentifier(this.CurrentToken));
             if (precedence > Precedence.Lambda)
             {
                 return false;
             }
 
+            if (this.CurrentToken.Kind == SyntaxKind.OpenBracketToken)
+            {
+                var resetPoint = this.GetResetPoint();
+
+                _ = ParseAttributeDeclarations();
+                bool result;
+
+                switch (CurrentToken.Kind)
+                {
+                    case SyntaxKind.StaticKeyword:
+                    case SyntaxKind.IdentifierToken:
+                        result = IsPossibleLambdaExpressionCore(precedence);
```

Comment:
i don't love the assymetry here with the check above (though i grant it is correct).  i would prefer another case of "== Static || IsTrueIdentifier" to make it clear tehse chunks are in sync, and aren't dependent on an internal detail of IsTrueIdentifier implying that this case is sufficient.